File: QM/src/ws/hoyland/qm/Basket.java
package ws.hoyland.qm;

/**
 * 生产者，消费者模式
 * 
 * @author hoy
 * 
 */
public class Basket {
	private byte count = 0;

	public synchronized void push() {
		try {
			while (count == 1) {
				this.wait();
			}
			this.notify();
		} catch (InterruptedException e) {
			e.printStackTrace();
		} catch (IllegalMonitorStateException e) {
			e.printStackTrace();
		}
		count = 1;
	}

	public synchronized void pop() {
		try {
			while (count == 0) {
				this.wait();
			}
			this.notify();
		} catch (InterruptedException e) {
			e.printStackTrace();
		} catch (IllegalMonitorStateException e) {
			e.printStackTrace();
		}
		count = 0;
	}
}

File: sszs-sln/sszs/Basket.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace ws.hoyland.util
{
    public class Basket
    {
        private byte count = 0;

        public void push()
        {
            lock (this)
            {
                try
                {
                    while (count == 1)
                    {
                        Monitor.Wait(this);
                    }
                    Monitor.PulseAll(this);
                }
                catch (Exception e)
                {
                    throw e;
                }
                count = 1;
            }
        }

        public void pop()
        {
            lock (this)
            {
                try
                {
                    while (count == 0)
                    {
                        Monitor.Wait(this);
                    }
                    Monitor.PulseAll(this);
                }
                catch (Exception e)
                {
                    throw e;
                }
                count = 0;
            }
        }
    }
}
